File: ActionUI/Helpers/DateHelper.swift
//
//  DateHelper.swift
//  ActionUI

import Foundation

/// Helper for parsing and formatting dates in ActionUI
struct DateHelper {
    /// Parses ISO 8601 date strings flexibly
    /// Supports: "2024-07-16", "2024-07-16T14:30:00Z", "2024-07-16T14:30:00+00:00", etc.
    /// - Parameter dateString: ISO 8601 formatted date string
    /// - Returns: Parsed Date, or nil if parsing fails
    /// - Note: Date-only strings (YYYY-MM-DD) are parsed to noon in the current timezone to avoid day shifts
    static func parseDate(from dateString: String) -> Date? {
        let formatter = ISO8601DateFormatter()
        
        // Try with date only (YYYY-MM-DD)
        // Parse in local timezone at noon to avoid day shifts across timezones.
        // Skip when a time component is present (contains 'T'): a full datetime
        // must keep its time, not be snapped to noon (the time-bearing branches
        // below handle it).
        if !dateString.contains("T") {
            formatter.formatOptions = [.withFullDate]
            formatter.timeZone = TimeZone.current
            if let date = formatter.date(from: dateString) {
                // Adjust to noon in the current timezone
                var components = Calendar.current.dateComponents([.year, .month, .day], from: date)
                components.hour = 12
                components.minute = 0
                components.second = 0
                if let noonDate = Calendar.current.date(from: components) {
                    return noonDate
                }
                return date
            }
        }

        // Try with date and time, no timezone designator (e.g.
        // "2024-07-16T09:05:00"). An offset-less datetime is wall-clock local
        // time, so parse it in the current timezone (the time-bearing DatePicker
        // serialization emits this form via formatDateTime).
        formatter.formatOptions = [.withFullDate, .withTime, .withColonSeparatorInTime]
        formatter.timeZone = TimeZone.current
        if let date = formatter.date(from: dateString) {
            return date
        }
        formatter.timeZone = nil
        
        // Try with date and time (no fractional seconds)
        formatter.formatOptions = [.withInternetDateTime]
        if let date = formatter.date(from: dateString) {
            return date
        }
        
        // Try with full date-time format first (with time and timezone)
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: dateString) {
            return date
        }

        // Try a bare local time "HH:mm" or "HH:mm:ss" - seed today's date with it.
        // Used by time-bearing DatePicker modes whose initial value is a clock time.
        if let timed = parseLocalTime(from: dateString) {
            return timed
        }

        return nil
    }

    /// Parses a bare local time string ("HH:mm" or "HH:mm:ss") into today's date
    /// at that time in the current timezone, or nil if it is not a bare time.
    static func parseLocalTime(from string: String) -> Date? {
        let trimmed = string.trimmingCharacters(in: .whitespaces)
        let parts = trimmed.split(separator: ":", omittingEmptySubsequences: false)
        guard parts.count == 2 || parts.count == 3 else { return nil }
        guard let hour = Int(parts[0]), let minute = Int(parts[1]) else { return nil }
        let second = parts.count == 3 ? (Int(parts[2]) ?? -1) : 0
        guard (0...23).contains(hour), (0...59).contains(minute), (0...59).contains(second) else { return nil }
        var components = Calendar.current.dateComponents([.year, .month, .day], from: Date())
        components.hour = hour
        components.minute = minute
        components.second = second
        return Calendar.current.date(from: components)
    }

    /// Formats a date as a full ISO 8601 calendar date "yyyy-MM-dd" (date-only).
    /// Mirrors the value-bridge serialization for date-only DatePickers.
    static func formatDate(from date: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]
        return formatter.string(from: date)
    }

    /// Formats a date as a full local ISO 8601 datetime "yyyy-MM-ddTHH:mm:ss"
    /// (no timezone designator), used by the time-bearing DatePicker modes so a
    /// host can read the wall-clock HH:mm. The components are taken in the
    /// current timezone to match what the user sees in the picker.
    static func formatDateTime(from date: Date) -> String {
        let c = Calendar.current.dateComponents([.year, .month, .day, .hour, .minute, .second], from: date)
        return String(format: "%04d-%02d-%02dT%02d:%02d:%02d",
                      c.year ?? 0, c.month ?? 0, c.day ?? 0,
                      c.hour ?? 0, c.minute ?? 0, c.second ?? 0)
    }
}

File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Helpers/DateHelper.kt
package com.abracode.actionui.Helpers

import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

/**
 * Date-string parsing/formatting for ActionUI Android. The Android counterpart
 * of Apple's `ActionUI/Helpers/DateHelper.swift` - the single place that turns a
 * JSON ISO 8601 date string into the value-bridge [LocalDate] and back, used by
 * the `DatePicker` element and the `DATE` branch of the value-string bridge in
 * `ActionUIModel`.
 *
 * The value bridge carries a date as a [java.time.LocalDate] (date-only,
 * timezone-free). This matches Apple's `.withFullDate` serialization (`DatePicker`
 * uses `displayedComponents: .date`) and sidesteps the day-shift-across-timezones
 * problem the Swift helper works around by parsing to local noon: a [LocalDate]
 * has no time-of-day to shift.
 *
 * [parseDate] accepts the same flexible ISO 8601 inputs as the Swift helper -
 * a bare full date (`2024-07-16`), or a date-time with or without an offset
 * (`2024-07-16T14:30:00Z`, `2024-07-16T14:30:00+01:00`, `2024-07-16T14:30:00`) -
 * and reduces any of them to the date component. Returns `null` on anything else.
 */
object DateHelper {

    /**
     * Parses a flexible ISO 8601 [dateString] to a date-only [LocalDate], or
     * `null` if no supported form matches. A date-time input is reduced to its
     * date component.
     */
    fun parseDate(dateString: String): LocalDate? {
        val trimmed = dateString.trim()
        if (trimmed.isEmpty()) return null
        return tryParse { LocalDate.parse(trimmed) }                                   // 2024-07-16
            ?: tryParse { OffsetDateTime.parse(trimmed).toLocalDate() }                // ...T..Z / +hh:mm
            ?: tryParse { Instant.parse(trimmed).atZone(ZoneOffset.UTC).toLocalDate() } // ...T..Z (instant)
            ?: tryParse { LocalDateTime.parse(trimmed).toLocalDate() }                 // ...T.. (no offset)
    }

    /** Formats [date] as a full ISO 8601 date (`yyyy-MM-dd`), matching Apple's `.withFullDate`. */
    fun formatDate(date: LocalDate): String = date.format(DateTimeFormatter.ISO_LOCAL_DATE)

    /**
     * Parses a flexible ISO 8601 [input] to a [LocalDateTime] (date + wall-clock
     * time), or `null` if no supported form matches. Used by the time-bearing
     * `DatePicker` modes (`hourAndMinute` / `dateAndTime`). Accepts:
     *   * a full local datetime (`2024-07-16T09:05:00`, `2024-07-16T09:05`);
     *   * an offset/instant datetime, reduced to its local date+time;
     *   * a bare time (`09:05`, `09:05:00`), seeded with today's date;
     *   * a bare date (`2024-07-16`), seeded at midnight.
     */
    fun parseDateTime(input: String): LocalDateTime? {
        val trimmed = input.trim()
        if (trimmed.isEmpty()) return null
        return tryParseDT { LocalDateTime.parse(trimmed) }                                    // ...T..:..(:..)
            ?: tryParseDT { OffsetDateTime.parse(trimmed).toLocalDateTime() }                 // ...T..Z / +hh:mm
            ?: tryParseDT { Instant.parse(trimmed).atZone(ZoneOffset.UTC).toLocalDateTime() } // ...T..Z (instant)
            ?: tryParseDT { LocalTime.parse(trimmed).atDate(LocalDate.now()) }                // 09:05(:..) -> today
            ?: tryParseDT { LocalDate.parse(trimmed).atStartOfDay() }                         // 2024-07-16 -> midnight
    }

    /**
     * Formats [dateTime] as a full local ISO 8601 datetime `yyyy-MM-ddTHH:mm:ss`
     * (no offset, seconds always present), the wire form the time-bearing
     * `DatePicker` modes emit so a host can read the wall-clock `HH:mm`.
     */
    fun formatDateTime(dateTime: LocalDateTime): String =
        dateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss"))

    private inline fun tryParse(block: () -> LocalDate): LocalDate? =
        try {
            block()
        } catch (e: DateTimeParseException) {
            null
        }

    private inline fun tryParseDT(block: () -> LocalDateTime): LocalDateTime? =
        try {
            block()
        } catch (e: DateTimeParseException) {
            null
        }
}
